apps/heft/src/pluginFramework/StaticFileSystemAdapter.ts

```diff
@@ -0,0 +1,234 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import type { ReaddirAsynchronousMethod, ReaddirSynchronousMethod } from '@nodelib/fs.scandir';
+import type { StatAsynchronousMethod, StatSynchronousMethod } from '@nodelib/fs.stat';
+import type { FileSystemAdapter } from 'fast-glob';
+import { Path } from '@rushstack/node-core-library';
+
+interface IVirtualFileSystemEntry {
+  name: string;
+  children?: Set<IVirtualFileSystemEntry>;
+}
+
+interface IReaddirOptions {
+  withFileTypes?: boolean;
+}
+
+/* eslint-disable @rushstack/no-new-null */
+type StatCallback = (error: NodeJS.ErrnoException | null, stats: fs.Stats) => void;
+type ReaddirStringCallback = (error: NodeJS.ErrnoException | null, files: string[]) => void;
+type ReaddirDirentCallback = (error: NodeJS.ErrnoException | null, files: fs.Dirent[]) => void;
+/* eslint-enable @rushstack/no-new-null */
+
+const IS_WINDOWS: boolean = process.platform === 'win32';
+
+/**
+ * A filesystem adapter for use with the "fast-glob" package. This adapter uses a static set of paths
+ * to provide a virtual filesystem.
+ *
+ * @remarks This adapter only implements methods required to allow for globbing. This means that it
+ * does not support returning true-to-disk file stats or dirent objects. Instead, the returned file
+ * stats and dirent objects only implement the `isDirectory` and `isFile` methods, which are
+ * required for filesystem traversal performed by the globber.
+ */
+export class StaticFileSystemAdapter implements FileSystemAdapter {
+  private _directoryMap: Map<string, IVirtualFileSystemEntry> = new Map<string, IVirtualFileSystemEntry>();
+
+  /** { @inheritdoc fs.lstat } */
+  public lstat: StatAsynchronousMethod = this._statInternal.bind(this);
+
+  /** { @inheritdoc fs.lstatSync } */
+  public lstatSync: StatSynchronousMethod = this._statSyncInternal.bind(this);
+
+  /** { @inheritdoc fs.stat } */
+  public stat: StatAsynchronousMethod = this._statInternal.bind(this);
+
+  /** { @inheritdoc fs.statSync } */
+  public statSync: StatSynchronousMethod = this._statSyncInternal.bind(this);
+
+  /** { @inheritdoc fs.readdir } */
+  public readdir: ReaddirAsynchronousMethod = this._readdirInternal.bind(this);
+
+  /** { @inheritdoc fs.readdirSync } */
+  public readdirSync: ReaddirSynchronousMethod = (this._readdirSyncInternal as ReaddirSynchronousMethod).bind(
+    this
+  );
```

Comment:
We often use the syntax:
```ts
  public lstat: typeof fs.lstat = (filePath: string, statCallback: StatCallback): void => {
    // Method body
  }
```
For autobound functions

Reply:
So I actually converted these to full-blown methods, and that ended up being a problem because fast-glob tries to substitute the missing FS functions with the native ones, and it does it using `Object.assign`, which ignores the methods on the prototype. So, I needed to make them properties again. When I did that, ESLint got all mad about methods being declared before the constructor. It seems that ESLint parses this style as a method, even though it's a property:
```typescript
public func: () => void = () => {}
```
However, by wrapping in braces and casting (which I need to do anyway for `readdirSync` due to the annoyingly complicated type), ESLint correctly interprets it as a public property. So I ended up going this route for all the methods.